File: frontend/lib/data/models/expense.dart
class Expense {
  final int? id;
  final String name;
  final String? description;
  final double amount;
  final String category;
  final DateTime date;

  Expense({
    this.id,
    required this.name,
    this.description,
    required this.amount,
    required this.category,
    required this.date,
  });

  factory Expense.fromMap(Map<String, dynamic> map) {
    return Expense(
      id: map['id'],
      name: map['name'],
      description: map['description'],
      amount: map['amount'].toDouble(),
      category: map['category'],
      date: map['date'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'description': description,
      'amount': amount,
      'category': category,
      'date': date,
    };
  }
}

File: backend/src/main/java/org/artso/budget_manager/expense/Expense.java
package org.artso.budget_manager.expense;

import com.fasterxml.jackson.annotation.JsonFormat;
import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.artso.budget_manager.auth.AppUser;
import org.artso.budget_manager.group.UserGroup;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Set;

@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Expense {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    @NotBlank
    private String name;
    private String description;
    @Min(0)
    private BigDecimal amount;
    private String category;
    @PastOrPresent
    @NotNull
    @JsonFormat(pattern = "yyyy-MM-dd")
    private LocalDate date;

    @ManyToOne
    @JoinColumn(name = "author_id")
    private AppUser author;

    @ManyToMany
    @JoinTable(
            name = "expense_shared_groups",
            joinColumns = @JoinColumn(name = "expense_id"),
            inverseJoinColumns = @JoinColumn(name = "group_id")
    )
    private Set<UserGroup> sharedGroups;
}
